src/ripple/app/misc/NegativeUNLVote.cpp

```diff
@@ -0,0 +1,354 @@
+        auto nUnlToDisable = prevLedger->nUnlToDisable();
+        auto nUnlToReEnable = prevLedger->nUnlToReEnable();
+        if (nUnlToDisable)
+            nUnlKeys.insert(*nUnlToDisable);
+        if (nUnlToReEnable)
+            nUnlKeys.erase(*nUnlToReEnable);
+
+        hash_set<NodeID> nUnlNodeIDs;
+        for (auto const& k : nUnlKeys)
+        {
+            auto nid = calcNodeID(k);
+            nUnlNodeIDs.emplace(nid);
+            if (nidToKeyMap.find(nid) == nidToKeyMap.end())
+            {
+                nidToKeyMap.emplace(nid, k);
+            }
+        }
+
+        auto const seq = prevLedger->info().seq + 1;
+        purgeNewValidators(seq);
+
+        std::vector<NodeID> toDisableCandidates;
+        std::vector<NodeID> toReEnableCandidates;
+        findAllCandidates(
+            unlNodeIDs,
+            nUnlNodeIDs,
+            scoreTable,
+            toDisableCandidates,
+            toReEnableCandidates);
+
+        if (!toDisableCandidates.empty())
+        {
+            auto n =
+                pickOneCandidate(prevLedger->info().hash, toDisableCandidates);
+            assert(nidToKeyMap.find(n) != nidToKeyMap.end());
+            addTx(seq, nidToKeyMap[n], true, initialSet);
+        }
+
+        if (!toReEnableCandidates.empty())
+        {
+            auto n =
+                pickOneCandidate(prevLedger->info().hash, toReEnableCandidates);
+            assert(nidToKeyMap.find(n) != nidToKeyMap.end());
+            addTx(seq, nidToKeyMap[n], false, initialSet);
+        }
+    }
+}
+
+void
+NegativeUNLVote::addTx(
+    LedgerIndex seq,
+    PublicKey const& vp,
+    bool disabling,
+    std::shared_ptr<SHAMap> const& initialSet)
+{
+    STTx nUnlTx(ttUNL_MODIFY, [&](auto& obj) {
+        obj.setFieldU8(sfUNLModifyDisabling, disabling ? 1 : 0);
+        obj.setFieldU32(sfLedgerSequence, seq);
+        obj.setFieldVL(sfUNLModifyValidator, vp.slice());
+    });
+
+    uint256 txID = nUnlTx.getTransactionID();
+    Serializer s;
+    nUnlTx.add(s);
+    auto tItem = std::make_shared<SHAMapItem>(txID, s.peekData());
+    if (!initialSet->addGiveItem(tItem, true, false))
+    {
+        JLOG(j_.warn()) << "N-UNL: ledger seq=" << seq
+                        << ", add ttUNL_MODIFY tx failed";
+    }
+    else
+    {
+        JLOG(j_.debug()) << "N-UNL: ledger seq=" << seq
+                         << ", add a ttUNL_MODIFY Tx with txID: " << txID
+                         << ", the validator to "
+                         << (disabling ? "disable: " : "re-enable: ") << vp;
+    }
+}
+
+NodeID
+NegativeUNLVote::pickOneCandidate(
+    uint256 randomPadData,
+    std::vector<NodeID>& candidates)
+{
+    assert(!candidates.empty());
+    static_assert(NodeID::bytes <= uint256::bytes);
+    NodeID randomPad = NodeID::fromVoid(randomPadData.data());
+    NodeID txNodeID = candidates[0];
+    for (int j = 1; j < candidates.size(); ++j)
+    {
+        if ((candidates[j] ^ randomPad) < (txNodeID ^ randomPad))
```

Comment:
We may want to cache the `txNodeID ^ randomPad` so it's not computed on every loop iteration. I suspect the optimizer will do this for us, but we may want to do this anyway. You're call. I'm OK with it as-is too.

Reply:
Since you are OK with it as-is, I did not change, because I think most likely candidates will have one element inside.